### PR TITLE
csvy_model

### PR DESCRIPTION
**Pull request by marxwillia**

### CODE REVIEW

File: tardis/io/tests/data/csvy_test_config.yml [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# Example YAML configuration for TARDIS
+tardis_config_version: v1.0
+
+supernova:
+  luminosity_requested: 8.770 log_lsun
+  time_explosion: 16 day
+  distance: 8.32 Mpc
+
+# standard atomic data base; get it from the tardis-refdata repository
+atom_data: /home/mew488/Research/TARDIS/tardis-refdata/atom_data/kurucz_cd23_chianti_H_He.h5
```
Comment (Youssef15015) [inline tardis/io/tests/data/csvy_test_config.yml]: Wouldn't it need to reference the tardis repo itself? As in: github.com/tardis-sn/tardis-refdata/blob/master/atom_data/kurucz_cd23_chianti_H_He.h5

Comment (marxwillia) [reply tardis/io/tests/data/csvy_test_config.yml]: Yes good point. Although this file was used for local testing and is not necessary for the pytests.

File: tardis/io/model_reader.py [hunk 2]
```diff
@@ -439,3 +439,48 @@ def read_csv_isotope_abundances(fname, delimiter='\s+', skip_columns=0,
                 z, mass_no), :] = df.loc[element_symbol_string].tolist()
 
     return abundance.index, abundance, isotope_abundance
+
+def parse_csvy_abundances(csvy_data):
```
Comment (wkerzendorf) [inline tardis/io/model_reader.py]: this looks nice

File: tardis/io/tests/data/csvy_test_config.yml [hunk 3]
```diff
@@ -0,0 +1,77 @@
+# Example YAML configuration for TARDIS
+tardis_config_version: v1.0
+
+supernova:
+  luminosity_requested: 8.770 log_lsun
+  time_explosion: 16 day
+  distance: 8.32 Mpc
```
Comment (wkerzendorf) [inline tardis/io/tests/data/csvy_test_config.yml]: this is likely deprecated - or at least should be.

File: tardis/io/tests/data/csvy_test_config.yml (hunk 1 above)
Comment (wkerzendorf) [reply tardis/io/tests/data/csvy_test_config.yml]: @Youssef15015 that won't work as tardis can not fetch online data. @marxwillia you can just use `kurucz_...` - TARDIS can download the necessary data.

File: tardis/io/tests/data/csvy_test_config.yml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+# standard atomic data base; get it from the tardis-refdata repository
+atom_data: /home/mew488/Research/TARDIS/tardis-refdata/atom_data/kurucz_cd23_chianti_H_He.h5
+
+csvy_model: csvy_full.csvy
+
+model:
```
Comment (wkerzendorf) [inline tardis/io/tests/data/csvy_test_config.yml]: how does that work? what takes precedent?

File: tardis/io/tests/data/csvy_test_config.yml [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      #w7_rho_0: 3e29 g/cm^3
+      #---------------
+
+  abundances:
+    type: file
+    filename: test_abund.dat
+    filetype: custom_composition
+
+plasma:
+  #initial_t_inner: 10000 K
```
Comment (wkerzendorf) [inline tardis/io/tests/data/csvy_test_config.yml]: same here

File: tardis/model/base.py [hunk 6]
```diff
@@ -6,7 +6,10 @@
 from tardis import constants
 
 from tardis.util.base import quantity_linspace
-from tardis.io.model_reader import read_density_file, read_abundances_file, read_uniform_abundances
+from tardis.io.parsers.csvy import load_csvy
+from tardis.io.model_reader import read_density_file, read_abundances_file, read_uniform_abundances, parse_csvy_abundances
```
Comment (wkerzendorf) [inline tardis/model/base.py]: more than 80 chars - I'm relatively sure

File: tardis/model/base.py [hunk 7]
```diff
@@ -383,3 +386,124 @@ def from_config(cls, config):
                    v_boundary_inner=structure.get('v_inner_boundary', None),
                    v_boundary_outer=structure.get('v_outer_boundary', None),
                    electron_densities=electron_densities)
+    @classmethod
+    def from_csvy(cls, config):
```
Comment (wkerzendorf) [inline tardis/model/base.py]: horrible written function 😉 - but I know that this is not from you. Should we clean it up - or defer to later?

File: tardis/model/density.py [hunk 8]
```diff
@@ -33,6 +33,61 @@ def calculate_density_at_time_of_simulation(self, time_explosion):
         return calculate_density_after_time(self.density_0, self.time_0,
                                             time_explosion).cgs
 
+    @classmethod
+    def from_csvy(cls, config, csvy_model_config):
+        """
+        Create a new HomologousDensity instance from a base Configuration object and a csvy model Configuration object.
```
Comment (wkerzendorf) [inline tardis/model/density.py]: more than 80 chars.

File: tardis/model/density.py [hunk 9]
```diff
@@ -33,6 +33,61 @@ def calculate_density_at_time_of_simulation(self, time_explosion):
...
+        ----------
+        config : tardis.io.config_reader.Configuration
+        csvy_model_config : tardis.io.config_reader.Configuration
+
+        Returns
+        -------
+        HomologousDensity
+
+        """
+        if hasattr(csvy_model_config, 'velocity'):
```
Comment (wkerzendorf) [inline tardis/model/density.py]: also horrible written - also not from you 😉

File: tardis/model/tests/data/config_csvy_full.yml [hunk 10]
```diff
@@ -0,0 +1,77 @@
+# Example YAML configuration for TARDIS
+tardis_config_version: v1.0
+
+supernova:
+  luminosity_requested: 8.770 log_lsun
+  time_explosion: 16 day
+  distance: 8.32 Mpc
+
+# standard atomic data base; get it from the tardis-refdata repository
+atom_data: /home/mew488/Research/TARDIS/tardis-refdata/atom_data/kurucz_cd23_chianti_H_He.h5
```
Comment (wkerzendorf) [inline tardis/model/tests/data/config_csvy_full.yml]: same problem.

File: tardis/model/tests/data/config_csvy_full_rad.yml [hunk 11]
```diff
@@ -0,0 +1,77 @@
+# Example YAML configuration for TARDIS
+tardis_config_version: v1.0
+
+supernova:
+  luminosity_requested: 8.770 log_lsun
+  time_explosion: 16 day
+  distance: 8.32 Mpc
+
+# standard atomic data base; get it from the tardis-refdata repository
+atom_data: /Users/marcwilliamson/Research/TARDIS/tardis-refdata/atom_data/kurucz_cd23_chianti_H_He.h5
```
Comment (wkerzendorf) [inline tardis/model/tests/data/config_csvy_full_rad.yml]: same problem

File: tardis/model/tests/data/csvy_full_rad.csvy [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+  fields:
+    -  name: velocity
+       unit: km/s
+       desc: velocities of shell outer bounderies.
+    -  name: density
+       unit: g/cm^3
+       desc: density of shell.
+    -  name: t_rad
+       unit: K
+       desc: radiative temperature.
```
Comment (wkerzendorf) [inline tardis/model/tests/data/csvy_full_rad.csvy]: that is optional, right?

File: tardis/model/tests/test_csvy_model.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+import tardis
+import os
+from astropy import units as u
+from tardis.io.config_reader import Configuration
+from tardis.model import Radial1DModel
+import pytest
+
+DATA_PATH = os.path.join(tardis.__path__[0],'model','tests','data')
+
+@pytest.fixture
```
Comment (wkerzendorf) [inline tardis/model/tests/test_csvy_model.py]: parametrize your fixtures - @chvogl can help

File: tardis/model/base.py (hunk 7 above)
Comment (marxwillia) [reply tardis/model/base.py]: I think let's defer that to later, and a different PR.  But I agree that it is not that well organized.  I will think about how to improve this.

File: tardis/model/base.py [hunk 14]
```diff
@@ -383,3 +386,124 @@ def from_config(cls, config):
...
+
+        """
+        if os.path.isabs(config.csvy_model):
+            csvy_model_fname = config.csvy_model
+        else:
+            csvy_model_fname = os.path.join(config.config_dirname,
+                                            config.csvy_model)
+        csvy_model_config, csvy_model_data = load_csvy(csvy_model_fname)
+        base_dir = os.path.abspath(os.path.dirname(__file__))
+        schema_dir = os.path.join(base_dir, '..', 'io', 'schemas')
```
Comment (wkerzendorf) [inline tardis/model/base.py]: that should also not live here.

File: tardis/io/model_reader.py [hunk 15]
```diff
@@ -439,3 +439,48 @@ def read_csv_isotope_abundances(fname, delimiter='\s+', skip_columns=0,
...
+    Returns
+    -------
+
+    index : ~np.ndarray
+    abundances : ~pandas.DataFrame
+    isotope_abundance : ~pandas.MultiIndex
+    """
+
+    abundance_col_names = [name for name in csvy_data.columns if nucname.iselement(name) or nucname.isnuclide(name)]
+    df = csvy_data.loc[:,abundance_col_names]
```
Comment (wkerzendorf) [inline tardis/io/model_reader.py]: pep 8 ;-)

File: tardis/model/tests/data/csvy_full_rad.csvy (hunk 12 above)
Comment (marxwillia) [reply tardis/model/tests/data/csvy_full_rad.csvy]: Yes.

File: tardis/io/tests/data/csvy_test_config.yml (hunk 3 above)
Comment (marxwillia) [reply tardis/io/tests/data/csvy_test_config.yml]: Yes the distance is depreciated.  I will remove it.

File: tardis/io/tests/data/csvy_test_config.yml (hunk 4 above)
Comment (marxwillia) [reply tardis/io/tests/data/csvy_test_config.yml]: How the model is loaded is handled by tardis/simulation/base.py in the from_config() class method. Currently, if the config has the 'csvy_model' keyword, then that is how the model is loaded.  I think this should actually be handled in the schema, where the user is only allowed to define either 'model' or 'csvy_model'

File: tardis/model/tests/test_csvy_model.py (hunk 13 above)
Comment (marxwillia) [reply tardis/model/tests/test_csvy_model.py]: Done.  Very helpful!

File: tardis/io/tests/data/csvy_test_config.yml (hunk 4 above)
Comment (marxwillia) [reply tardis/io/tests/data/csvy_test_config.yml]: raise error if both are defined.

File: tardis/model/base.py (hunk 14 above)
Comment (marxwillia) [reply tardis/model/base.py]: validation should be done in load_csvy.

**Comment by wkerzendorf** [comment]
can you rebase to the master branch - because there are some changes from youssef in there.

**Pull request merged**